packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -975,15 +978,24 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
         this.updateDocumentDirtyState(newState);
     }
 
-    public setConnectionState(connected: boolean, clientId?: string) {
+    public async setConnectionState(connected: boolean, clientId?: string) {
```

Comment:
I think you are inviting can of worms by making it async. We can lose connection and get another connection while previous async invocation is in flight. That will create a lot of problems (even this code is not correct in such conditions, same for all the functionality called from here).